presto-parquet/src/main/java/com/facebook/presto/parquet/reader/AbstractColumnReader.java

```diff
@@ -166,18 +186,87 @@ private void processValues(int valuesToRead, Consumer<Void> valueConsumer)
                 valueConsumer.accept(null);
                 valueCount++;
                 if (valueCount == remainingValueCountInPage) {
-                    updateValueCounts(valueCount);
+                    updateValueCounts(valueCount, 0);
+                    if (!readNextPage()) {
+                        return;
+                    }
+                    valueCount = 0;
+                }
+                repetitionLevel = repetitionReader.readLevel();
+                definitionLevel = definitionReader.readLevel();
+            }
+            while (repetitionLevel != 0);
+        }
+        updateValueCounts(valueCount, 0);
+    }
+
+    /**
+     * When filtering using column indexes we might skip reading some pages for different columns. Because the rows are
+     * not aligned between the pages of the different columns it might be required to skip some values. The values (and the
+     * related rl and dl) are skipped based on the iterator of the required row indexes and the first row index of each
+     * page.
+     * For example:
+     *
+     * rows   col1   col2   col3
+     *      ┌──────┬──────┬──────┐
+     *   0  │  p0  │      │      │
+     *      ╞══════╡  p0  │  p0  │
+     *  20  │ p1(X)│------│------│
+     *      ╞══════╪══════╡      │
+     *  40  │ p2(X)│      │------│
+     *      ╞══════╡ p1(X)╞══════╡
+     *  60  │ p3(X)│      │------│
+     *      ╞══════╪══════╡      │
+     *  80  │  p4  │      │  p1  │
+     *      ╞══════╡  p2  │      │
+     * 100  │  p5  │      │      │
+     *      └──────┴──────┴──────┘
+     *
+     * The pages 1, 2, 3 in col1 are skipped so we have to skip the rows [20, 79]. Because page 1 in col2 contains values
+     * only for the rows [40, 79] we skip this entire page as well. To synchronize the row reading we have to skip the
+     * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+     * skip values while reading page0 and page1 for col3.
+     */
+    private void processValuesSync(int valuesToRead, Consumer<Void> valueConsumer)
```

Comment:
shall we have a better name for this function?
seems we are leveraging pageIterators to skip reading some pages, right? An idea implementation would be merge `processValuesSync` and `processValues` into one function, `processValues`, with a boolean parameter, skipPages?

Reply:
Any suggestion for the name? 

I agree better to merge. How about we open a Jira/Issue to do it later. The reason for it is as the initial version, keeping them separate is less error-pone at this point because when this feature is not enabled by default all the existing queries are not impacted.

Reply:
I am fine with the Presto coding style if you have a strong opinion on it. Let me change it.